src/conv/invokers/impl_gemm_dynamic.cpp

```diff
@@ -73,45 +75,198 @@ float CallImplicitGemmDynamic(const miopen::Handle& handle,
+    opArgs.emplace_back(wei);
+    opArgs.emplace_back(src);
+    opArgs.emplace_back(hi);
+    opArgs.emplace_back(wi);
+    opArgs.emplace_back(n);
+    opArgs.emplace_back(k);
+    opArgs.emplace_back(c);
+    opArgs.emplace_back(ho);
+    opArgs.emplace_back(wo);
+    opArgs.emplace_back(stride_h);
+    opArgs.emplace_back(stride_w);
+    opArgs.emplace_back(dilation_h);
+    opArgs.emplace_back(dilation_w);
+    opArgs.emplace_back(pad_h);
+    opArgs.emplace_back(pad_w);
+    opArgs.emplace_back(y);
+    opArgs.emplace_back(x);
+    opArgs.emplace_back(dtile_iy);
+    opArgs.emplace_back(dtile_ix);
+    opArgs.emplace_back(dtile_dy);
+    opArgs.emplace_back(dtile_dx);
+    opArgs.emplace_back(dtile_y);
+    opArgs.emplace_back(dtile_x);
+    opArgs.emplace_back(dtile_h);
+    opArgs.emplace_back(dtile_w);
+    opArgs.emplace_back(dslice_y);
+    opArgs.emplace_back(dslice_x);
+    opArgs.emplace_back(dslice_h);
+    opArgs.emplace_back(dslice_w);
+    opArgs.emplace_back(dslice_h_left);
+    opArgs.emplace_back(dslice_w_left);
+    opArgs.emplace_back(__pack0);
+
+    for(int gemm_id = 0; gemm_id < num_of_gemms; gemm_id++)
     {
-        return [ctx](const std::vector<Kernel>& kernels) {
-            return [=](const Handle& handle, const boost::any& primitive_parameters) {
-                const auto data_ctx = boost::any_cast<conv::DataInvokeParams>(primitive_parameters);
-                const auto& tensors = data_ctx.tensors;
-                auto kernel         = handle.Run(kernels[0]);
-                if(kernel.GetName().find("igemm_v4r1_dynamic") == 0 ||
-                   kernel.GetName().find("igemm_v4r1_1x1_dynamic") == 0)
+        int _dtile_iy          = gemm_id / x_tilda;
+        int _dtile_ix          = gemm_id % x_tilda;
+        int _y_dot_slice       = (_dtile_iy + 1) * y_dot <= y ? y_dot : y % y_dot;
+        int _x_dot_slice       = (_dtile_ix + 1) * x_dot <= x ? x_dot : x % x_dot;
+        int _gemm_k            = k * _y_dot_slice * _x_dot_slice;
+        bool is_gemm_not_empty = _gemm_k > 0;
+        opArgs[18]             = OpKernelArg(_dtile_iy);
+        opArgs[19]             = OpKernelArg(_dtile_ix);
+        opArgs[26]             = OpKernelArg(_y_dot_slice);
+        opArgs[27]             = OpKernelArg(_x_dot_slice);
+        if(is_gemm_not_empty)
+            kernel(opArgs);
+    }
+
+    if(handle.IsProfilingEnabled())
+        elapsed += handle.GetKernelTime();
+    return elapsed;
+}
+
+InvokerFactory MakeImplGemmDynamicForwardInvokerFactory(const ConvolutionContext& ctx)
+{
+    return [ctx](const std::vector<Kernel>& kernels) {
+        return [=](const Handle& handle, const boost::any& primitive_parameters) {
+            const auto data_ctx = boost::any_cast<conv::DataInvokeParams>(primitive_parameters);
+            const auto& tensors = data_ctx.tensors;
+            auto kernel         = handle.Run(kernels[0]);
+            if(kernel.GetName().find("igemm_v4r1_dynamic") == 0 ||
```

Comment:
Invokers should not check on kernel names. It is much preferred to check on some data, cache some bool value, or have separate invokers for separate paths. Checking on the kernel name is a sign for that invoker should be split to multiple. This not the case for this particular one for now as I see, but in the future, you should make separate invokers for different execution paths. I am really sorry I had to make igemm shared invoker as it is, but time was the limiting factor and my puny experience with igemm solvers didn't allow me to split the invoking code at that moment, offloading that to future developments.
Avoiding the huge runtime blocks with spaghettified invocation code is part of the reason behind the invokers' system.

Reply:
@DrizztDoUrden Oh now I understand. Invokers are intent to serve one or group of kernels that have the same input argument right? If kernel arg or path is different, should separate into different invokers.
Just to make sure, in dynamic-igemm, or some other algorithms, that have multiple kernels to serve as one solver (like kernel opt for ho/wo is even, or opt for 1x1 filter, etc...). If that multiple kernel lives in the same asm or hip source file, the kernel name must be different. So in this case, the logic to choose which kernel to use should live in solver, not invoker, am I correct?

Reply:
cc @shaojiewang please notice

Reply:
kernel name related check in invoker is removed

Reply:
Yeah, exactly.

Reply:
> cc @shaojiewang please notice

Yes, got it.